velox/exec/HashTable.cpp

```diff
@@ -568,6 +576,153 @@ bool HashTable<ignoreNullKeys>::insertBatch(
       }
     }
   }
+  if (hashMode_ == HashMode::kNormalizedKey && initNormalizedKeys) {
+    for (auto i = 0; i < rows.size(); ++i) {
+      RowContainer::normalizedKey(rows[i]) = hashes[i];
+      hashes[i] = mixNormalizedKey(hashes[i], sizeBits_);
+    }
+  }
+  return true;
+}
+
+namespace {
+template <typename Source>
+void syncWork(std::vector<std::shared_ptr<Source>>& work) {
+  for (auto& item : work) {
+    try {
+      item->move();
+    } catch (const std::exception& e) {
+      // Ignore errors.
+    }
+  }
+}
+} // namespace
+template <bool ignoreNullKeys>
+void HashTable<ignoreNullKeys>::parallelJoinBuild() {
+  int32_t numPartitions = 1 + otherTables_.size();
+  VELOX_CHECK_GT(
+      size_ / numPartitions,
+      160,
+      "Less than 160 entries per partition for parallel build");
+  partitionBounds_.resize(numPartitions + 1);
+  // Pad the tail of partitionBounds_ to max int.
+  std::fill(
+      partitionBounds_.begin(),
+      partitionBounds_.begin() + partitionBounds_.capacity(),
+      std::numeric_limits<int32_t>::max());
+  for (auto i = 0; i < numPartitions; ++i) {
+    // The bounds are rounded up to cache line size.
+    partitionBounds_[i] = bits::roundUp((size_ / numPartitions) * i, 64);
+  }
+  partitionBounds_.back() = size_;
+  std::vector<std::shared_ptr<AsyncSource<bool>>> partitionWork;
+  std::vector<std::shared_ptr<AsyncSource<bool>>> buildWork;
+  auto sync = folly::makeGuard([&]() {
+    syncWork(partitionWork);
+    syncWork(buildWork);
+  });
+  for (auto i = 0; i < numPartitions; ++i) {
+    auto table = i == 0 ? this : otherTables_[i - 1].get();
+    partitionWork.push_back(
+        std::make_shared<AsyncSource<bool>>([this, table, numPartitions]() {
+          partitionRows(*table, numPartitions);
+          return std::make_unique<bool>(true);
+        }));
+    buildExecutor_->add([work = partitionWork.back()]() { work->prepare(); });
+  }
+  syncWork(partitionWork);
+  std::vector<std::vector<char*>> overflow(numPartitions);
+  for (auto i = 0; i < numPartitions; ++i) {
+    buildWork.push_back(
+        std::make_shared<AsyncSource<bool>>([i, &overflow, this]() {
+          buildJoinPartition(i, overflow[i]);
+          return std::make_unique<bool>(true);
+        }));
+    buildExecutor_->add([work = buildWork.back()]() { work->prepare(); });
+  }
+  syncWork(buildWork);
+  raw_vector<uint64_t> hashes;
+  for (auto& overflows : overflow) {
+    hashes.resize(overflows.size());
+    hashRows(
+        folly::Range<char**>(overflows.data(), overflows.size()),
+        false,
+        hashes);
+    insertForJoin(
+        overflows.data(), hashes.data(), overflows.size(), 0, size_, nullptr);
+  }
+}
+
+namespace {
+int32_t
+findPartition(int32_t index, const int32_t* bounds, int32_t numPartitions) {
```

Comment:
Should we fall back to binary search if `numPartitions` is large?

Reply:
Yeah, we could tune this later based on benchmark result.